src/mcp/server/fastmcp/server.py

```diff
@@ -822,12 +828,165 @@ async def read_resource(self, uri: str | AnyUrl) -> Iterable[ReadResourceContent
+
+        interaction_id = interaction_id or str(uuid.uuid4())
+        interaction_data = {
+            "message": {"type": "text", "text": message},
+            "schema": schema,
+        }
+
+        # Track progress if supported
+        progress_available = bool(
+            client_capabilities.userInteraction
+            and client_capabilities.userInteraction.progress
+        )
+
+        result = await self.request_context.session.create_user_interaction(
+            interaction_id=interaction_id,
+            interaction_type="prompt",
+            interaction_data=interaction_data,
+            progress_available=progress_available,
+            related_request_id=self.request_id,
+        )
+
+        return result.content or {}
+
+    async def request_user_agent_interaction(
+        self,
+        url: str,
+        message: str | None = None,
+        interaction_id: str | None = None,
+    ) -> None:
+        """Request the user to interact with a URL via a user agent (browser).
+
+        This method is suitable for sensitive interactions that should happen directly
+        between the user and a third-party service, such as authentication flows.
+
+        Args:
+            url: The URL the user should navigate to
+            message: Optional explanatory message about the interaction to show the user
+            interaction_id: Optional custom ID for the interaction
+                (one will be generated if not provided)
+
+        Raises:
+            ValueError: If user agent interaction is not supported by the client
+        """
+        # Verify client capability
+        client_capabilities = (
+            self.request_context.session.client_params.capabilities
+            if self.request_context.session.client_params
+            else None
+        )
+
+        if (
+            not client_capabilities
+            or not client_capabilities.userInteraction
+            or (
+                client_capabilities.userInteraction.types
+                and "ua" not in client_capabilities.userInteraction.types
+            )
+        ):
+            raise ValueError("Client does not support user agent interaction type")
+
+        interaction_id = interaction_id or str(uuid.uuid4())
+        interaction_data = {"url": url}
+
+        if message:
+            interaction_data["message"] = {"type": "text", "text": message}
+
+        # Track progress if supported
+        progress_available = bool(
+            client_capabilities.userInteraction
+            and client_capabilities.userInteraction.progress
+        )
+
+        await self.request_context.session.create_user_interaction(
+            interaction_id=interaction_id,
+            interaction_type="ua",
+            interaction_data=interaction_data,
+            progress_available=progress_available,
+            related_request_id=self.request_id,
+        )
+
+    def require_user_agent_interaction(
```

Comment:
I intentionally did not add a developer facing function for requiring an interaction for type `prompt`. If someone really wanted to do that, then they could always use the `require_user_interaction` method in `server/session.py`